File: 백준/Silver/16564. 히오스 프로게이머/히오스 프로게이머.java
import java.io.*;
import java.util.*;

class Main{
    static int n;
    static long k;
    static long bs(){
        long l=0,r=1000000001;
        while(l<=r){
            long m=(l+r)/2,cn=n*m;
            if(cn<k) l=m+1;
            else if(cn>k) r=m-1;
            else return m;
        }
        return r;
    }
    public static void main(String[] args)throws Exception{
        BufferedReader I=new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st=new StringTokenizer(I.readLine());
        n=Integer.parseInt(st.nextToken());
        k=Integer.parseInt(st.nextToken());
        long[] nb=new long[n],num=new long[n+1];

        for(int i=0;i<n;i++) nb[i]=Integer.parseInt(I.readLine());
        Arrays.sort(nb);
        for(int i=0;i<n;i++) num[i+1]=num[i]+nb[i];
        boolean bl=true;
        
        for(int i=0;i<n;i++){
            if(k<nb[i]*i-num[i]){
                k+=num[i];
                n=i;
                bl=false;
                break;
            }
        }
        if(bl) k+=num[n];
        System.out.print(bs());
    }
}

File: 백준/Silver/16564. 히오스 프로게이머/히오스 프로게이머.py
import sys
def input(): return sys.stdin.readline().rstrip()
def bs():
    l,r=0,1000000001
    while l<=r:
        m=(l+r)//2
        cn=n*m
        if cn<k:
            l=m+1
        elif cn>k:
            r=m-1
        else: return m
    return r

n,k=map(int,input().split())
nb,num=[0]*n,[0]*(n+1)

for i in range(n):
    nb[i]=int(input())
    
nb.sort()
   
for i in range(n):
    num[i+1]+=num[i]+nb[i]
    
bl=True
for i in range(n):
    if k<nb[i]*i-num[i]:
        k+=num[i]
        n=i
        bl=False
        break

if bl: k+=num[n]
print(bs())